### ISSUE TITLE
Allow renaming a folder name without destroying the folder

### ISSUE THREAD

**Issue by LudovicEmoAirbus** [terraform-google-modules/terraform-google-folders#41]
### TL;DR

Note: We have originally opened issue 40, which has been closed because similar to issue 38.
However issue 38 is also closed with no change delivered, stating that this is due to a terraform constraint.
However we reject this reason and have coming with a proposed solution.

Request:
If we rename a previously created folder through _names_ variable, terraform will destroy the folder and recreate a new one, as the folder name is used as a for_each key.

Note 1: the workaround to manually remove the folder from the state and delete the folder is not acceptable, the goal is to automate with Terraform and avoiding manual actions.
Note 2: Also the potential workaround to add the new folder name as a new value in the list and keep the old name as it is is not acceptable as we do not want to keep a second folder, we need a renaming.

### Terraform Resources

```markdown
names = ["previous_name"]
changed to:
names = ["new_name"]

-------------------------------
Plan result:

# module.module1.google_folder.folders["previous_name"] will be destroyed
[...]
# module.module1.google_folder.folders["new_name"] will be created
[...]
```

### Detailed design

```markdown
The variables needs to be adapted to isolate the requested folder display name from the for_each resource key to allow renaming.

For example:
instead of flat names variable, use a list such as:
     names= [
                      {folder_key = "my_folder_1", folder_name= "requested_display_name"},
    ]

and for folders resource, replace toset with something similar to:
resource "google_folder" "folders" {
  for_each = {for folder in var.names:  folder.folder_key => folder}
  display_name = "${local.prefix}${each.value.folder_name}"

This way, when we change folder_name later to "new_name", we do not change folder_key, the state key is unchanged and the folder is renamed as expected.
```

_Originally posted by @LudovicEmoAirbus in https://github.com/terraform-google-modules/terraform-google-folders/issues/38#issuecomment-1119760634_

**Comment by lemoo5sg** [comment]
Hi,
Another alternative might be to keep names variable as it is and to use count instead of for_each, to use indexes as tf state resources keys instead of the names, however on one side the impact on folder_admin_roles_map_data  & owners resources has to be assessed and the other side it will create an implicit ordering of the names, new folders will have to be added at the end to have no impact on the existing ones.

```markdown
resource "google_folder" "folders" {
  count= length(var.names)

  display_name = "${local.prefix}${var.names[count.index]}"
  parent       = var.parent
}
```

**Comment by bharathkkb** [comment]
@lemoo5sg Are [moved blocks](https://www.terraform.io/language/modules/develop/refactoring) another alternative you could consider? The first approach maybe possible to implement but it would be a breaking change and some added complexity.

**Comment by brettcurtis** [comment]
I'm playing around with some code to get by this myself, and it still feels pretty ugly :D - NOTE: I'm not using this module but what I'm wondering is if anyone has any links to good examples of folks using this resource in a reasonably DRY way? At this point, for me, it's more around subfolders and just copy-pasting resources with different parent keys.

**Comment by Ludovic-Emo-Pyl-Tech** [comment]
Hi,
Could someone please confirm if this issue could been addressed or is already resolved through another case?
I’m currently avoiding the use of this module and relying on my own implementation due to this concern.

**Comment by red8888** [comment]
Has development of this module ceased? Its not a terribly complex module and the solution is pretty straight forward. 

Do you accept PRs? Maybe I or someone can open a PR to fix this?

I realized I accidentally opened a dup little while ago, but so is everyone else lol. If I had known this limitation I would not have used this module.
